Parallel export

clslgrnc opened the pull request:
Following our brief exchange on mastodon, here is a complete parallel export code, with functions and call graph.

With 5 workers I get a 2x speedup on  a 1MB binary and 100% match on callgraph and functions from a regular export.

- Each job is assigned some of the functions to export.
- all resulting db are merged
- a final job exports the remaining data

I refactored all SQL insertions related to functions in order to easily switch between:
- sequential export: rows are inserted sequentially
- parallel export: ensure `rowid % nbr_jobs = job_id` in order to avoid collisions when merging

run with: `IDADIR=<path-to-ida> ./diaphora_parallel_export.py <path-to-target-binary>`

Two potential improvements that remain to be done:
- [x] distribute address ranges to workers without relaunching IDA (needs a communication channel with the ida scripts)
- [ ] use a threadsafe db in order to parallelize merges

Sequential export still seems to work.

clslgrnc committed:
Refactor SQL queries related to functions

Extract all SQL queries related to functions into a dedicated class.
It will then be easier to alter all of them at once in the context
of parallel export of functions data.

If further SQL queries related to functions are necessary, they
should be added to this class.

clslgrnc committed:
Parallel export

`diaphora_parallel_export.py` performs the following actions:
- first, `idat64` is used to perform the auto-analysis and get an
  idb;
- then, a queue manager with two queues is created:
  - `job_queue` is used to send jobs to the workers;
  - `report_queue` is used by workers to report jobs done and
    termination.
- a number of workers are launched: they copy the idb and launch
  `idat64` with a script that retrieves the queues and await for
  instructions;
- jobs and kill switches are sent;
- as soon as possible (when all jobs are performed or in-progress),
  the generated SQLite databases, containing functions information,
  are merged, resulting in a database with all functions data but
  no program data;
- `idat64` is used one last time to retrieve those program data.

In order to avoid collisions while merging databases, each worker
only use database indices such that:
`index % nbr_workers == worker_id`.

In order to choose the functions to analyze, a worker just divide
the sorted list of functions into as many parts as the total
number of jobs, and process the nth part (where n is the job id).

joxeankoret commented:
I will not integrate it into the main branch for obvious reasons, but I will leave this PR here so people can use it if they like.

clslgrnc committed:
Refactor do_export

Extract the list of functions exported by `do_export` into the
method `filter_functions`. This will allow a parallel version of
`do_export` to build a generator of functions to export on the fly
while receiving instruction from a main process.

Additionally, parallel export will not always export functions in
a predictable order. In the context of crashed exports, instead of
relying on the address of the last function inserted into the
database, all addresses added are retrieved and `self._funcs_cache`
is restored.

clslgrnc committed:
Refactor do_export logging

Extract argument of `replace_wait_box` into a variable, so that it
can be used as a log line instead of a message box update.

And a slight cosmetic change to the frequency of updates.

clslgrnc committed:
// export: keep db connection open

Keep the connection to the main database open.
Not sure if it really is faster.

clslgrnc commented:
I can think of several obvious reasons not to integrate this into the main branch :smiley:

If you could integrate any (or all) of the first three commits (87a57af4c66f33fda54d50f444d9288a0e85f0f9, 14503cf473d9ee20abce621806968865ccaaba01, d389cb076cb543bd00218ba438cd50d441877f83) it would make maintaining my fork easier though. If you are interested, let me know which commits would be acceptable so that I can prepare a PR (if not, fair enough).

For anyone interested please post any issue about `diaphora_parallel_export.py` over there: https://github.com/clslgrnc/diaphora-parallel-export/issues

clslgrnc commented:
From [#264 ](https://github.com/joxeankoret/diaphora/issues/264#issuecomment-1642659150)
> Not a feature so much, but optimizing the export speed would be nice. I have an arm64 binary that takes over 2 and a half hours to export at the moment. `[Diaphora: Wed Jul 19 21:07:13 2023] Database exported, time taken: 2:43:39.113402.`

@Myles1 the `diaphora_parallel_export.py` script in this PR might be able to speed this up.
It launches several instances of idat64 on copies of the target idb, thus RAM might be a limiting factor.
Let me know if you have any question.

> It's just so nice to be able to automate function matching in this way. You've made an enormously helpful program.

Indeed, thank you @joxeankoret

arm64eabi commented:
hello @clslgrnc, do you have plans to solve merge conflicts? I'd like to use it (and maybe more people)

clslgrnc commented:
I do not want to make any promises, and even if I fix the merge conflicts I'll probably not take the time to properly test it.
If I do, I'll let you know.

Do not expect an amazing speed up though. In my initial testing I needed 5 workers to get a 2x speed up on a small sample. Moreover, since each worker is a full IDA instance you'll need a lot of memory to properly speed up the analysis of a large binary.

Chapoly1305 commented:
> I will not integrate it into the main branch for obvious reasons, but I will leave this PR here so people can use it if they like.

Sorry to ask after years from original post, what was initial reason not merging or revise the code base? I had a program ran for a long while and this PR is quite tempting.
